gnupg-pkcs11-scd-proxy/gnupg-pkcs11-scd-proxy.1
.\"
.\" Copyright (c) 2006-2017 Alon Bar-Lev <alon.barlev@gmail.com>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"
.\"     o Redistributions of source code must retain the above copyright notice,
.\"       this list of conditions and the following disclaimer.
.\"     o Redistributions in binary form must reproduce the above copyright
.\"       notice, this list of conditions and the following disclaimer in the
.\"       documentation and/or other materials provided with the distribution.
.\"     o Neither the name of the <ORGANIZATION> nor the names of its
.\"       contributors may be used to endorse or promote products derived from
.\"       this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd October 15, 2017
.Os POSIX-compatible
.Dt gnupg-pkcs11-scd-proxy 1
.Sh NAME
.Nm gnupg-pkcs11-scd-proxy
.Nd GnuPG-compatible smart-card proxy daemon
.Nm gnupg-pkcs11-scd-proxy-server
.Nd GnuPG-compatible smart-card proxy server daemon
.Sh SYNOPSIS
.Nm gnupg-pkcs11-scd-proxy
.Op --multi-server
.Op --socket Ar file
.Op --verbose
.Op --log-file Ar file
.Op --help
.Nm gnupg-pkcs11-scd-proxy-server
.Op --socket Ar file
.Op --socket-group Ar group
.Op --scd Ar file
.Op --scd-config Ar file
.Op --verbose
.Op --log-file Ar file
.Op --help
.Sh DESCRIPTION
.Nm gnupg-pkcs11-scd-proxy
is a drop-in replacement for the smart-card daemon (scd) shipped with the
next-generation GnuPG (gnupg-2). The daemon is a proxy into
.Nm gnupg-pkcs11-scd-proxy-server
which can be run within different security context to load the
.Nm gnupg-pkcs11-scd
smart-card daemon.
.Pp
The communications between the proxy and the server is unix socket based.
The assuan socket is created as world readable but with a specific user
ACL, so that only initiating user can connect to the daemon.
.Pp
The recommended version of gnupg is 2.1 since in this version the
.Nm gnupg-pkcs11-scd
configuration does not specify any specific key information and can be
used as generic for all users.
.Pp
The following options are available:
.Bl -tag -width "AA"
.It --multi-server
Run in multi-server mode (foreground). In addition to communicating over
stdin/stdout, the server also opens an additional listening UNIX socket.
.It --socket Ar file
Socket name, default should be sufficient.
.It --socket-group Ar group
A custom group to set for the proxy server socket, this may serve first
level of access control.
.It --scd Ar file
The smartcard daemon location, default should be sufficient.
.It --scd-config Ar file
The smartcard daemon configuration, required parameter.
.It --verbose
Be verbose while running.
.It --no-detach
Do not detach from console (useful for debugging purposes).
.It --log-file Ar file
Output log to
.Ar file .
.It --help
Print help information.
.El
.Pp
When the daemon receives any of the SIGHUP, SIGTERM and SIGINT signals,
it cleans up and exits.
.Sh SEE ALSO
.Xr gnupg-pkcs11-scd 1
.Rs
.%T "gnupg-pkcs11 Home Page"
.%O http://gnupg-pkcs11.sourceforge.net
.Re
.Sh AUTHORS AND COPYRIGHT
Copyright (c) 2006-2007 Zeljko Vrba <zvrba@globalnet.hr>
.Pp
Copyright (c) 2006-2017 Alon Bar-Lev <alon.barlev@gmail.com>
.Pp
All rights reserved.
.Pp
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


ChangeLog
gnupg-pkcs11-scd
Copyright (c) 2006-2007 Zeljko Vrba <zvrba@globalnet.hr>
Copyright (c) 2006-2017 Alon Bar-Lev <alon.barlev@gmail.com>

????-??-?? - Version 0.9.2

 * Support openssl-1.1, thanks Thorsten Alteholz, W. Michael Petullo.

2017-09-26 - Version 0.9.1

 * Support unix domain socket credentials on FreeBSD.
 * Introduce GNUPG_PKCS11_SOCKETDIR to instruct where sockets are created.
 * Make proxy systemd service work again per change of systemd behavior.

2017-08-25 - Version 0.9.0

 * Avoid dup of stdin/stdout so that the terminate assuan hack operational
   again.
 * Introduce gnupg-pkcs11-scd-proxy to allow isolation of the PKCS#11
   provider.
 * Lots of cleanups.

2017-07-15 - Version 0.8.0

 * Support multiple tokens via serial numbers by hashing token id into
   serial number.
   Implementation changes the card serial number yet again, executing
   gpg --card-status should resync.

2017-04-18 - Version 0.7.6

 * Add --homedir parameter.
 * Rework serial responses for gnupg-2.1.19.

2017-03-01 - Version 0.7.5

 * Fix issue with decrypting padded data, thanks to smunaut.
 * Catchup with gnupg-2.1 changes which caused inability to support
   both gpg and gpgsm. Implementation had to change card serial
   number, as a result current keys of gpg will look for the
   previous serial card.
   emulate-openpgpg option is obsoleted and removed.

   ACTION REQUIRED
   in order to assign new card serial number to existing keys.
   backup your ~/.gnupg.
   delete all PKCS#11 secret keys using:
       gpg --delete-secret-keys $KEY then
   Then refresh keys using:
       gpg --card-edit
   In <gnupg-2.1.19 the keys should be re-generated using:
       admin
       generate
   Do not replace keys!
   gpg will learn the private keys of the new card and attach to
   the existing public keys.
 * Support gnupg-2.1 features of using existing keys, keys
   should not be explicitly specified in configuration file
   any more.

2017-01-18 - Version 0.7.4

 * Fix gpg change in serialno attribute.
 * Sync with gnupg-2.1, thanks to Moritz Bechler.

2011-07-30 -- Version 0.7.3

 * Use assuan_sock_init, bug#3382372.

2011-04-09 -- Version 0.7.2

 * Some cleanups, thanks to Timo Schulz.
 * Sync hashing algorithms for OpenPGP.

2011-03-16 -- Version 0.7.1

 * Sync with gnupg-2.0.17.

2010-05-28 -- Version 0.7.0

 * Add timegm() replacement, thanks to Lars Silvén.
 * Sync with gnupg-2.0.14 commands.
 * Need assuan-2.

2008-01-19 -- Version 0.06

 * Fixed uninitializied gnutls datum, thank to Volodya
   Vaynshteyn for his help.

 * Sync with gnupg-2.0.8 assuan usage.

 * Some minor fixups.

2007-10-05 -- Version 0.05

 * Fixed a change in gnupg protocol for openpgp
   emulation, thanks to Nick Kossifidis for testing.

 * Added key timestamp on OpenPGP key generation,
   thanks to Daniel Benoy for reporting and testing.

2007-06-13 -- Version 0.04

 * Build system fixups.

 * RPM packaging is available, thank to Eddy Nigg.

 * Debian packaging is available, thank to Sandro Wefel.

 * Hacks to support for OpenPGP smartcard emulation,
   please refer the man page.

2007-01-05 -- Version 0.03

 * Fixed some Intel C compiler warnings.

 * pkcs11-helper is now external dependency.

 * Decrypt support unwrap now, thanks for
   Christoph Neerfeld help.

 * Configuration file sign-mode was renamed to
   private-mask.

2006-10-20 -- Version 0.02

 * Added more characters to serialization exclude chars.
   People may need to re-learn cards.

 * Fixed configure GPG_HOME typeo, thanks to Bob Dunlop.

 * Unlike gpg-agent, gnupg-pkcs11-scd supports more than
   one token available at the same time. In order to make
   gpg-agent happy, gnupg-pkcs11-scd always returns the
   same card serial number to gpg-agent.
   When unavailable token is requested, gnupg-pkcs11-scd
   will use NEEDPIN callback in order to ask for the
   requested token.
   When and if gpg-agent will support more than one serial
   number or NEEDTOKEN callback, this behavior will be
   modified.

 * Works as single threaded win32 server. Implementing
   multi-threaded server forces copy of win32 socket handling
   from gnupg, and it is GPLed.
   When libassuan will handle sockets we will be able to
   implement it.

 * Added system wide configuration file, thanks to Bub Dunlop.

 * Some more minor fixes.

2006-10-11 -- Version 0.01

 * Initial release.


gnupg-pkcs11-scd-proxy/gnupg-pkcs11-scd-proxy.c
/*
 * Copyright (c) 2006-2017 Alon Bar-Lev <alon.barlev@gmail.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *     o Redistributions of source code must retain the above copyright notice,
 *       this list of conditions and the following disclaimer.
 *     o Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     o Neither the name of the <ORGANIZATION> nor the names of its
 *       contributors may be used to endorse or promote products derived from
 *       this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <errno.h>
#include <getopt.h>
#include <poll.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>

static volatile int s_stop = 0;

static RETSIGTYPE sigterm(int signo) {
	(void)signo;
	s_stop = 1;
#if RETSIGTYPE != void
	return 0
#endif
}

static void usage(char *name) {
	printf (
		(
"%s %s\n"
"\n"
"Copyright (c) 2006-2017 Alon Bar-Lev <alon.barlev@gmail.com>\n"
"This program comes with ABSOLUTELY NO WARRANTY.\n"
"This is free software, and you are welcome to redistribute it\n"
"under certain conditions. See the file COPYING for details.\n"
"\n"
"Syntax: %s [options]\n"
"Smartcard daemon for GnuPG\n"
"\n"
"Options:\n"
" \n"
"     --multi-server        run in multi server mode (foreground)\n"
"     --homedir             specify home directory\n"
"     --socket=FILE         use this socket\n"
" -v, --verbose             verbose\n"
"     --log-file            use a log file for the server\n"
"     --help                print this information\n"
		),
		PACKAGE,
		PACKAGE_VERSION,
		name
	);
}

int main(int argc, char *argv[]) {
	struct sockaddr_un addr;
	int fd = -1;
	int ret = 1;
	long on = 1;
	char *socket_name = CONFIG_PROXY_SOCKET;

	typedef struct fds_s {
		int fd;
		char *name;
		int peer;
		char buffer[1024];
		size_t buffer_n;
	} fds_t;

	fds_t fds[] = {
		{ -1, "outgoing.out", 2, {0}, 0},
		{ 0, "incoming.in", 0, {0}, 0},
		{ 1, "incoming.out", 1, {0}, 0},
	};
	int fds_n = sizeof(fds) / sizeof(fds[0]);
	int disconnect = 0;

	enum {
		OPT_MUTLI_SERVER,
		OPT_HOMEDIR,
		OPT_SOCKET,
		OPT_VERBOSE,
		OPT_LOG_FILE,
		OPT_VERSION,
		OPT_HELP
	};

	static struct option long_options[] = {
		{ "multi-server", no_argument, NULL, OPT_MUTLI_SERVER },
		{ "homedir", required_argument, NULL, OPT_HOMEDIR },
		{ "socket", required_argument, NULL, OPT_SOCKET },
		{ "verbose", no_argument, NULL, OPT_VERBOSE },
		{ "log-file", required_argument, NULL, OPT_LOG_FILE },
		{ "version", no_argument, NULL, OPT_VERSION },
		{ "help", no_argument, NULL, OPT_HELP },
		{ NULL, 0, NULL, 0 }
	};
	int opt;

	while ((opt = getopt_long (argc, argv, "v", long_options, NULL)) != -1) {
		switch (opt) {
			case OPT_MUTLI_SERVER:
			break;
			case OPT_HOMEDIR:
			break;
			case OPT_SOCKET:
				socket_name = optarg;
			break;
			case OPT_VERBOSE:
			case 'v':
			break;
			case OPT_LOG_FILE:
			break;
			case OPT_VERSION:
				printf (
					"%s %s\n"
					"\n"
					"Copyright (c) 2006-2017 Alon Bar-Lev <alon.barlev@gmail.com>\n"
					"\n"
					"This is free software; see the source for copying conditions.\n"
					"There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n",
					PACKAGE,
					PACKAGE_VERSION
				);
				exit (0);
			break;
			case OPT_HELP:
				usage(argv[0]);
				exit(0);
			break;
			default:
				fprintf(stderr, "invalid usage\n");
				exit(1);
			break;
		}
	}

	signal(SIGPIPE, SIG_IGN);
	signal(SIGTERM, sigterm);
	signal(SIGINT, sigterm);

	memset(&addr, 0, sizeof(addr));
	addr.sun_family = AF_UNIX;

	if (strlen (socket_name) + 1 >= sizeof (addr.sun_path)) {
		fprintf(stderr, "Socket '%s' too long, expected %ld\n", socket_name, (long)sizeof (addr.sun_path));
		goto cleanup;
	}
	strcpy(addr.sun_path, socket_name);

	if ((fd = socket(AF_UNIX, SOCK_STREAM, 0)) == -1) {
		perror("socket");
		goto cleanup;
	}

	if (ioctl(fd, FIONBIO, &on) == -1) {
		perror("ioctl sock");
		goto cleanup;
	}

	if (connect(fd, (struct sockaddr *) &addr, sizeof(struct sockaddr_un)) == -1) {
		fprintf(stderr, "Cannot connect '%s': %s\n", socket_name, strerror(errno));
		goto cleanup;
	}

	fds[0].fd = fd;
	while (!s_stop && !disconnect) {
		struct pollfd pollfds[fds_n];
		int i;

		memset(&pollfds, 0, sizeof(pollfds));

		for (i = 0; i < fds_n; i++) {
			fds_t *peer = &fds[fds[i].peer];
			pollfds[i].fd = fds[i].fd;
			if (peer->buffer_n < sizeof(peer->buffer)) {
				pollfds[i].events |= POLLIN;
			}
			if (fds[i].buffer_n > 0) {
				pollfds[i].events |= POLLOUT;
			}
		}

		if (poll(pollfds, sizeof(pollfds) / sizeof(struct pollfd), -1) == -1) {
			if (errno != EINTR && errno != EAGAIN) {
				perror("poll");
				goto cleanup;
			}
			continue;
		}

		for (i = 0; i < fds_n && !disconnect; i++) {
			if ((pollfds[i].revents & POLLHUP) != 0) {
				disconnect = 1;
			}
			if ((pollfds[i].revents & POLLERR) != 0) {
				fprintf(stderr, "error %s\n", fds[i].name);
				goto cleanup;
			}
		}

		for (i = 0; i < fds_n && !disconnect; i++) {
			if ((pollfds[i].revents & POLLIN) != 0) {
				fds_t *peer = &fds[fds[i].peer];
				int n;
				if ((n = read(fds[i].fd, peer->buffer + peer->buffer_n, sizeof(peer->buffer) - peer->buffer_n)) == -1) {
					fprintf(stderr, "error %s read\n", fds[i].name);
					goto cleanup;
				}
				if (n == 0) {
					disconnect = 1;
				}
				peer->buffer_n += n;
			}

			if ((pollfds[i].revents & POLLOUT) != 0) {
				int n;
				if ((n = write(fds[i].fd, fds[i].buffer, fds[i].buffer_n)) == -1) {
					fprintf(stderr, "error %s write\n", fds[i].name);
					goto cleanup;
				}
				fds[i].buffer_n -= n;
				memmove(fds[i].buffer, fds[i].buffer + n, fds[i].buffer_n);
			}
		}
	}

	ret = 0;

cleanup:
	if (fd != -1) {
		close(fd);
	}

	return ret;
}
